### ISSUE TITLE
the issue of frequency

### ISSUE THREAD

**Issue by a2821952** [yxyang/locomotion_simulation#2]
Thank you for your help, I try to detect the frequency of robot, and I found that the frequency is hard to reach 300HZ, and I'm wondering which computer on A1 you are using for MPC, Tx2 in the front part? or Upboard in the back?

**Comment by yxyang** [comment]
I used neither TX2 nor the Upboard. Both were too slow for my code (though if you convert everything to C++ it should work).

I got it to work on a Nvidia Jetson Agx Xavier at ~330Hz. A decent intel NUC could reach 400-500Hz. A intel 10th-gen i9 CPU could reach 1000Hz. I chose the Jetson Xavier cuz it could be powered using on-board power supply.

**Comment by a2821952** [comment]
Thank you for your explain!

**Comment by ShaohangXu** [comment]
Hi Yang @yxyang. Thanks for your code sharing.

I tested the code on the real robot with TX2. It worked well in sim but I did not get good results on the real robot.

I wrote a stand controller which sends initial motor angles. Besides, there is a one-second trot gait controlled by your mpc-based controller. I attached two tested videos in [Baidu Yun.](https://pan.baidu.com/s/1jJLdfyQu7gR21jKiufEOMQ). The password is 0h7a

One problem is that the first control step will finish after ~0.2s when the robot reset. This causes that the robot is in protection mode, which will brake the motors. I have to pull it up myself in case the robot falls down. Sometimes other control step would exteed the safe frequency (33Hz) too. Have you met this problem?


Another problem is that the robot stamped too heavily and almost fell when troting. I have to pull it up to protect it. The control frequency is ~150Hz, which is less than 300Hz you recommended. Will the frequency affect this so much? If it runs in ~300Hz, how steady it will be?

The official version of A1 has only TX2. May I ask how you change it to Agx Xavier? Is there any problem when installing it?

You said if turned to C++, the code would run faster. May I ask how to change it? Will it be possible using ROS? Or just pure C++ projects using the Unitree C++ SDK？

By the way, may I ask whether you have trained any RL-based controller through the a1 gym env?
